ibis/backends/duckdb/registry.py

```diff
@@ -74,36 +75,65 @@ def _timestamp_from_unix(t, op):
         raise ValueError(f"`{unit}` unit is not supported!")
 
 
+class struct_pack(GenericFunction):
+    def __init__(
+        self, mapping: Mapping[str, Any], *, type: sa.types.TypeEngine
+    ) -> None:
+        super().__init__()
+        self.mapping = mapping
+        self.type = type
+
+
+@compiles(struct_pack, "duckdb")
+def compiles_struct_pack(element, compiler, **kw):
+    args = ", ".join(
+        "{key} := {value}".format(key=key, value=compiler.process(value, **kw))
+        for key, value in element.mapping.items()
+    )
+    return f"struct_pack({args})"
+
+
 def _literal(t, op):
     dtype = op.output_dtype
     sqla_type = t.get_sqla_type(dtype)
+
     value = op.value
     if dtype.is_interval():
         return sa.literal_column(f"INTERVAL '{value} {dtype.resolution}'")
     elif dtype.is_set() or dtype.is_array():
-        return sa.cast(sa.func.list_value(*value), sqla_type)
-    elif isinstance(value, np.ndarray):
-        return sa.cast(sa.func.list_value(*value.tolist()), sqla_type)
-    elif isinstance(value, (numbers.Real, np.floating)) and np.isnan(value):
-        return sa.cast(sa.literal("NaN"), sqla_type)
-    elif isinstance(value, collections.abc.Mapping):
-        if dtype.is_struct():
-            placeholders = ", ".join(
-                f"{key} := :v{i}" for i, key in enumerate(value.keys())
-            )
-            text = sa.text(f"struct_pack({placeholders})")
-            bound_text = text.bindparams(
-                *(sa.bindparam(f"v{i:d}", val) for i, val in enumerate(value.values()))
-            )
-            name = op.name if isinstance(op, ops.Named) else "tmp"
-            params = {name: t.get_sqla_type(dtype)}
-            return bound_text.columns(**params).scalar_subquery()
+        return sa.cast(
+            sa.func.list_value(
+                *(value.tolist() if isinstance(value, np.ndarray) else value)
+            ),
+            sqla_type,
+        )
+    elif dtype.is_floating() and not np.isfinite(value):
```

Comment:
Could you move `np.isfinite()` check inside to handle all floating literals here (returning for finite values too)?

Reply:
Possibly, let me check.

Reply:
Done.